Feat/adapter

solomon-negusse opened the pull request:
This adds new `rest` dataset connector to GFW Data API.  Related PRs in other repos:

- documentation: https://github.com/resource-watch/doc-api/pull/190
- dataset: https://github.com/resource-watch/dataset/pull/132
- infrastructure: https://github.com/resource-watch/api-infrastructure/pull/103
- sql-compatibility-test: https://github.com/resource-watch/sql-compatibility-test/pull/1

tiagojsag reviewed (changes requested):
hey @solomon-negusse 

Overall I think this is a good start, but there are a few things that will need to be improved for this new MS to be production-ready. There are a number of things that need to be modified, so to keep things easier to manage, I'll list a few of them now, and I'll do follow up review(s) as needed once these changes are in place.

In no particular order:
- Your dockerfile uses node 12, but the rest of the code mentions node 16
- You have a mix of old and new env vars used by the RW API integration MS. Please update the lib to the latest version, and check its README for details on what's actually needed, and remove everything else
- There are multiple references to other MSs from which you copy-pasted code.
- There are multiple references to Control Tower
- Different places of your code reference different HTTP ports on which the service runs. This is technically fine, but it helps to use the same port everywhere, to avoid runtime/test/CI issues.
- the `app/microservice` folder is not needed, and can be deleted.
- TSlint has been deprecated, use eslint + ts plugin instead
- Your k8s deployment is called "gfw" which IMO is too generic and prone to cause confusion

solomon-negusse commented:
Thanks for the review @tiagojsag! I have addressed your comments.
Sorry for the control tower inconsistencies - I started  with older version of `rw-api-microservice-node` initially.

solomon-negusse commented:
Thanks for the review @tiagojsag! I have addressed all but the Control Tower part of your comments. There are references to it in the README as I used it in local dev. Do you suggest removing the references completely and then update when localstack is operational or indicate that it's a dependency for local dev only?

tiagojsag commented:
thank @solomon-negusse I'll re-review these tomorrow. I would remove any and all references to CT. As for localstack, that's a bridge we'll cross when we get there - but I don't think it makes sense to document something now when we dont know when it will be rolled out.

solomon-negusse commented:
That sounds good, thanks. I was actually not asking if we should add localstack doc now - it was whether we should wait to remove control tower reference until it's fully replaced with localstack for local dev. Sorry for the confusion.. removing CT shortly.

tiagojsag commented on app/src/errors/datasetNotFound.error.js:
```diff
@@ -0,0 +1,12 @@
+class DatasetNotFound extends Error {
```
convert to ts

tiagojsag commented on app/src/routes/api/v1/gfw.router.ts:
```diff
@@ -0,0 +1,248 @@
+import GfwService from 'services/gfw.service';
+import ErrorSerializer from 'serializers/errorSerializer';
+
+const API_VERSION: string = 'v1';
```
you are using this constant with multiple endpoints, meaning it will do more harm than good. I suggest removing it, and just using '/v1' in the different strings where it' currently being used.

tiagojsag commented on app/src/routes/api/v1/gfw.router.ts:
```diff
@@ -0,0 +1,248 @@
+            body: {
+                dataset: {
+                    datasetUrl: url
+                        .replace('/gfw', '')
```
replace with `.replace('/gfw/', '/')` and make the equivalent change on the line below

solomon-negusse replied:
Below is an example result of the existing code which looks right. Am I missing something as to why the change you suggest is needed?

```javascript
const datasetUrl = '/api/v1/gfw/query/1639574869912?sql=SELECT%20%2A%20from%20DATA%20LIMIT%202'
const resultUrl = datasetUrl.replace('/gfw', '').replace('/api/v1', '')
resultUrl
> '/query/1639574869912?sql=SELECT%20%2A%20from%20DATA%20LIMIT%202"
```

tiagojsag commented on app/src/routes/api/v1/gfw.router.ts:
```diff
@@ -0,0 +1,248 @@
+                json: true,
+            });
+        } catch (e) {
+            await RWAPIMicroservice.requestToMicroservice({
```
what happens if this call to an external MS fails?

solomon-negusse replied:
Now catching that and raising proper error for that scenario

tiagojsag commented on app/src/routes/api/v1/gfw.router.ts:
```diff
@@ -0,0 +1,248 @@
+            options.method = 'POST';
+        }
+    } else {
+        logger.debug('Obtaining sql from featureService');
```
test coverage shows that this else clause is not covered by tests. Is this needed, and used/supported by the GFW api?

solomon-negusse replied:
Removed feature service logic as it's not supported

tiagojsag commented on app/src/serializers/user.serializer.ts:
```diff
@@ -0,0 +1,37 @@
+import { IUser } from 'services/okta.interfaces';
```
AFAICT this whole class is dead code

tiagojsag commented on app/src/services/dataset.service.ts:
```diff
@@ -0,0 +1,38 @@
+import { RWAPIMicroservice } from 'rw-api-microservice-node';
+import logger from 'logger';
+
+// @ts-ignore
```
fix the underlying cause, and remove this ts-ignore

tiagojsag commented on app/src/services/dataset.service.ts:
```diff
@@ -0,0 +1,38 @@
+// @ts-ignore
+import DatasetNotFound from 'errors/datasetNotFound.error';
+
+const API_VERSION: string = 'v1';
```
this is used once, [KISS](https://en.wikipedia.org/wiki/KISS_principle) says remove it

tiagojsag commented on app/src/services/gfw.service.ts:
```diff
@@ -0,0 +1,74 @@
+        }
+    }
+
+    static async executeQuery(
```
this method and the following can probably be DRYed up

solomon-negusse replied:
Done.. not sure why it's not labeled as outdated.

tiagojsag commented on app/test/e2e/fields.spec.ts:
```diff
@@ -0,0 +1,355 @@
+
+chai.should();
+
+const fields: Record<string, any>[] = [
```
this probably makes more sense in the constants file

tiagojsag commented on app/test/e2e/query-get.spec.ts:
```diff
@@ -0,0 +1,123 @@
+        const { datasetUrl, application } = body.dataset;
+        application.should.deep.equal(['your', 'apps']);
+        datasetUrl.should.equal(
+            `/query/${timestamp}?sql=${encodeURI(sql).replace('*', '%2A')}`,
```
this `.replace()` call feels evil, as there are more characters that encodeURI does not escape. not a big deal, look into a replacement if you feel like it

solomon-negusse replied:
Will keep an eye out of that usage. Leaving it as here as it's being used for fixed known string

tiagojsag commented on app/src/services/gfw.service.ts:
```diff
@@ -0,0 +1,74 @@
+            const result: Record<string, any> = await axios.get(reqUrl);
+            return result.data;
+        } catch (err) {
+            logger.error('Error obtaining fields', err);
```
this is probably redundant, as the core error handler will log the error too

tiagojsag reviewed (changes requested):
Left a few comments on the code. Besides those:
- your tests are making actual calls to the GFW API, which IMO should not happen
- you have multiple invalid calls to `createMockGetDataset` with a single argument on your tests, when the actual function definition requires 2 arguments
- you have unnecessary imports of functions 

I think once these are tackled this should be good to merge and test live

solomon-negusse commented:
Thanks for the review @tiagojsag.  Linting is in better state now so future reviews should be less cumbersome.
I'll get this merged to dev and test aws deployment if all looks good.

tiagojsag commented on app/src/serializers/errorSerializer.ts:
```diff
@@ -0,0 +1,5 @@
+export default class ErrorSerializer {
```
for consistency, rename this file to match the others

tiagojsag reviewed (changes requested):
Just one minor change left, otherwise LGTM
